ES reliability

emord opened the pull request:
@benrudolph @czue @proteusvacuum 

Mentioned this in the  meeting today. It's currently only informational, but I imagine a world where this could resave cases/applications/whatever else.

Ran it on softlayer and it completed more quickly than I expected. enikshay-test is missing quite  a few forms/cases from ES which is concerning.

emord committed:
better use of ES for # of cases

emord committed:
Quick implementation of number form/cases in domain

emord committed:
Add option to get # forms/cases from primary DB

emord committed:
first pass at es reliability script

emord committed:
don't include app builds

emord committed:
cleanup and add domain diff in csv

emord committed:
start docs deep dive

emord committed:
add deep dive

emord committed:
sharing is caring

emord committed:
more sharing, more caring

emord committed:
lint

emord committed:
fix

emord committed:
silly mistake

emord commented on corehq/apps/domain/calculations.py:
```diff
@@ -74,7 +75,9 @@ def active_mobile_users(domain, *args):
 
 
-def cases(domain, *args):
+def cases(domain, primary_db=False, *args):
```
These two methods aren't used right now (they were earlier in the PR). keeping htem here becasue I think they will fit in well for proactive alerting (to hopefully come soon)

emord committed:
dimagimon

emord commented on corehq/form_processor/interfaces/dbaccessors.py:
```diff
@@ -130,6 +130,9 @@ def form_exists(self, form_id):
     def get_all_form_ids_in_domain(self, doc_type='XFormInstance'):
         return self.db_accessor.get_form_ids_in_domain_by_type(self.domain, doc_type)
 
+    def get_number_of_forms_in_domain(self):
```
This will preferably use count in the future

emord replied:
Couch domains now use count/reduce

emord committed:
final cleanup

benrudolph commented:
sweet, this is cool. my one thing i'm still not sure about is that we actually don't expect cases/forms/etc to be exactly equal in couch and es at a given point in time. we expect that t - 60mins (or some interval) that es and couch will be the same (where t is now). if we were to run this on prod, there'd be a large number of false positives. pretty much any domain that is in the process of submitting a form. i know it makes the problem way harder, but otherwise i think it'd be too noisy to do on a daily basis. still great for debugging a domain though

emord commented:
@benrudolph yeah. I think that the number could have false positives, but its easy to filter out small differences in excel or something later, and look to see how many domains have large differences. enikshay on india had ~100 in sql, but not es

emord committed:
Active domains by default

benrudolph commented on corehq/apps/hqadmin/management/commands/check_docs_in_es.py:
```diff
@@ -0,0 +1,136 @@
+        ])
+
+        case_ids_primary = set(CaseAccessors(domain).get_case_ids_in_domain())
+        form_ids_primary = set(FormAccessors(domain).get_all_form_ids_in_domain())
```
not sure how to get around this but this may OOM. maybe if it's an iterator you could incrementally write to a file

emord replied:
Yeah. I'm running into some issues with big domains https://github.com/dimagi/commcare-hq/pull/14259/commits/68e461bf9c8ffe3ec00c53d97b072ffbcc9d6d62 :/ I don't think it's OOM, but it's not a clear error

emord committed:
Handle errors more gracefully

There's currently a really really long error that can span hundreds of
lines, so throwing away for now

emord committed:
Don't return deleted cases by default

emord committed:
Include inactive users

emord committed:
do smart counting of couch forms

emord committed:
be smarter about not getting all case ids

emord committed:
fix bug

emord committed:
weird

emord committed:
better handling of big couch case records

czue commented:
took a quick look and this seems great. agree we'll need to tweak it when we try to use it as an active monitoring thing.

emord committed:
make form queries more similar

emord committed:
Use reduce

emord committed:
fix tests

emord commented on corehq/apps/es/cases.py:
```diff
@@ -21,6 +21,9 @@
 class CaseES(HQESQuery):
     index = 'cases'
+    default_filters = {
+        'is_commcare_case': filters.doc_type("CommCareCase"),
```
I'm not sure if this is the "right" thing to do. Currently anything that uses `CaseES` is returning all cases including ones that have been deleted. That doesn't seem like what we should be doing, and it's inconsistent with what couch/sql do by default.

czue replied:
the pillow should remove soft deletions. see [code](https://github.com/dimagi/commcare-hq/blob/master/corehq/ex-submodules/pillowtop/processors/elastic.py#L32) and [test](https://github.com/dimagi/commcare-hq/blob/master/testapps/test_pillowtop/tests/test_case_pillow.py#L47)

are there lots of `CommCareCase-Deleted` in the index?

czue replied:
see my other comment, but i would have thought deleted cases would be removed from the index

emord replied:
Hmm. It does appear like that's the case, but I ran into a couple of deleted cases when I was running this command. I don't have a great idea on why that is though. Possibly another bug or inconsistency in pillow top?

emord commented on corehq/form_processor/backends/sql/dbaccessors.py:
```diff
@@ -311,6 +311,10 @@ def get_forms_by_type(domain, type_, limit, recent_first=False):
 
     @staticmethod
+    def get_number_of_forms_in_domain(domain, doc_type='XFormInstance'):
+        return len(FormAccessorSQL.get_form_ids_in_domain_by_type(domain, doc_type))
```
These don't use `count`, but that's not really necessary as there's not any huge sql domains at the moment

czue replied:
my guess is we might want to address this before we add any automatic running of this

emord reviewed (commented):
dumped more context in http://manage.dimagi.com/default.asp?244248 for the next interrupt team

czue commented:
build is failing
```
======================================================================
FAIL: test_case_pillow (testapps.test_pillowtop.tests.test_case_pillow.CasePillowTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 264, in inner
    return helper(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 237, in __call__
    call_with_settings(fn_with_pre_and_post, run_config.settings, args, kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 252, in call_with_settings
    fn(*args, **kwargs)
  File "/mnt/commcare-hq-ro/corehq/util/test_utils.py", line 233, in fn_with_pre_and_post
    self.fn(*args, **kwargs)
  File "/mnt/commcare-hq-ro/testapps/test_pillowtop/tests/test_case_pillow.py", line 40, in test_case_pillow
    self.assertEqual(1, results.total)
AssertionError: 1 != 0
```

emord commented:
superseded by #14654
